google/cloud/internal/timer_queue.cc

```diff
@@ -46,80 +56,83 @@ void TimerQueue::Shutdown() {
   std::unique_lock<std::mutex> const lk(mu_);
   shutdown_ = true;
   cv_.notify_all();
+  cv_follower_.notify_all();
 }
 
-// Timers added via Schedule should be managed by one or more threads that
-// call Service. Calls to Service only return after Shutdown has been called.
 void TimerQueue::Service() {
-  while (true) {
-    std::unique_lock<std::mutex> lk(mu_);
-    auto const until = NextExpiration(lk);
-    // TODO(#10521): Predicate condition ne < until may need to be reworked.
-    auto predicate = [this, until, &lk] {
-      auto const ne = NextExpiration(lk);
-      return shutdown_ || ne < until ||
-             HasExpiredTimer(lk, std::chrono::system_clock::now());
+  // It is easier to understand this if you first ignore shutdown and empty
+  // timer queues.
+  //
+  // The threads play two different roles:
+  // - A single thread at a time plays the leader role.
+  // - All other threads are followers and block on cv_follower_.
+  //
+  // Once a timer expires the leader thread relinquishes its role and wakes up
+  // one follower thread to become the new leader. Only after doing so it
+  // runs the code to expire the timer.
+  //
+  // This is all complicated by shutdown. Basically all thread needs to wake
+  // up when the TQ is shutdown and one of them will expire all the timers.
```

Comment:
My version used all the threads to cancel the outstanding timers, which I thought was a feature.  You might consider it.

Reply:
FWIW, your version was holding the lock through the `while(!timers_.empty())` loop, so effectively only one thread was doing the work.  But I should have noticed the intent, and I like this idea.  Done and added a test.

Reply:
> FWIW, your version was holding the lock through the while(!timers_.empty()) loop

And FWIW, I'm not sure which version you might have been looking at, but it dropped/reobtained the lock during the `Expire()` helper.